libqrexec/toml.c

```diff
@@ -0,0 +1,275 @@
+                LOG(ERROR, "%s:%zu: Invalid character '%c' at start of key", config_full_path, lineno, current_line[0]);
+            } else {
+                LOG(ERROR, "%s:%zu: Invalid byte 0x%x at start of key", config_full_path, lineno, current_line[0]);
+            }
+            goto bad;
+        }
+        unsigned char *key_cursor = (unsigned char *)current_line;
+        do {
+            key_cursor++;
+        } while (qubes_is_key_byte(key_cursor[0]));
+        int const key_len = key_cursor - (unsigned char *)current_line;
+        while (qubes_isspace(key_cursor[0]))
+            key_cursor++;
+        if (key_cursor[0] != '=') {
+            LOG(ERROR, "%s:%zu: Missing '=' after key", config_full_path, lineno);
+            goto bad;
+        }
+        do {
+            key_cursor++;
+        } while (qubes_isspace(key_cursor[0]));
+        current_line[key_len] = '\0';
+        union toml_data value;
+        enum toml_type ty = parse_toml_value(config_full_path, lineno, key_cursor, &value);
+
+        if (strcmp(current_line, "wait-for-session") == 0) {
+            if (seen_wait_for_session) {
+                LOG(ERROR, "%s:%zu: Key '%s' appears more than once", config_full_path, lineno, current_line);
+                goto bad;
+            }
+            seen_wait_for_session = true;
+            if (ty == TOML_TYPE_BOOL) {
+                *wait_for_session = (int)value.boolean;
+                continue;
+            } else if (ty == TOML_TYPE_INTEGER) {
+                if (value.integer < 2) {
+                    *wait_for_session = (int)value.integer;
+                    continue;
+                }
+                LOG(ERROR, "Integer value %llu used when a boolean was expected", value.integer);
+            } else if (ty == TOML_TYPE_STRING) {
+                LOG(ERROR, "String value '%s' not valid for 'wait-for-session'", value.string);
+                free(value.string);
+                value.string = NULL;
+            }
+        } else if (strcmp(current_line, "user") == 0) {
+            if (seen_user) {
+                LOG(ERROR, "%s:%zu: Key '%s' appears more than once", config_full_path, lineno, current_line);
+                goto bad;
+            }
+            seen_user = true;
+            char *bad_type;
+            switch (ty) {
+            case TOML_TYPE_INVALID:
+                goto bad;
+            case TOML_TYPE_BOOL:
+                bad_type = "Boolean";
+                break;
+            case TOML_TYPE_INTEGER:
+                bad_type = "Integer";
+                break;
+            case TOML_TYPE_STRING:
+                *user = value.string;
+                continue;
+            default:
+                abort();
+            }
+            LOG(ERROR, "%s:%zu: %s not valid for user name or user ID", config_full_path, lineno, bad_type);
+        } else {
+            LOG(ERROR, "%s:%zu: Unsupported key %s", config_full_path, lineno, current_line);
```

Comment:
TBH, I'm not sure if unknown settings should fail the call, it will make future extensions harder since often the same qrexec service can be used across many qubes versions. Take this PR as an example - a service that wants to support both old and new version could use the `user` feature, but in its absence fallback to re-exec itself via sudo.